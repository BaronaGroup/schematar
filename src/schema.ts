import ObjectId from './object-id'
import Complex from './complex'

export type DefaultContext = 'mongoose' | 'typescript' | 'jsonschema'

export interface Schema {
    fields: SchemaFields
}

type SimpleType = Symbol | Complex | typeof String | typeof Number | typeof Boolean | typeof Date | typeof Object
type IndexOptions = false | true | 'unique' | 'unique-sparse'

export type PlainType = SimpleType | [FieldInfo | SimpleType]

export type Field = FieldInfo | PlainType

export interface FieldInfo {
    type: PlainType
    index?: IndexOptions
    format?: any
    optional?: boolean
    presentIn?: Array<string>
    optionalIn?: Array<string>
    enum?: string[]
    mongooseDefault?: any
    mongooseRef?: string
}

export interface SchemaFields {
    [name: string]: Field
}
