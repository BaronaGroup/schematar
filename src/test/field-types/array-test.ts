import {Schema} from '../../schema'
import {testJSONSchema, testMongooseField, testTypescriptInterface} from '../test-utils'
import Complex from '../../complex'

describe('array-test', function() {
  describe('array of simple values', function() {
    const schema: Schema = {fields: {
      field: [Number]
    }}

    it('json-schema', testJSONSchema(schema, data => {
      expect(data).toMatchSnapshot()
    }))

    it('mongoose', testMongooseField(schema, 'field', field => {
      expect(field).toEqual({type: [{type: Number}]})
    }))

    it('typescript', testTypescriptInterface(schema, tsInterface => {
      expect(tsInterface).toMatchSnapshot()
    }))
  })

  describe('array of simple values defined with options', function() {
    const schema: Schema = {fields: {
      field: [{type: Number, index: true}]
    }}

    it('json-schema', testJSONSchema(schema, data => {
      expect(data).toMatchSnapshot()
    }))

    it('mongoose', testMongooseField(schema, 'field', field => {
      expect(field).toEqual({type: [{type: Number, index: true}]})
    }))

    it('typescript', testTypescriptInterface(schema, tsInterface => {
      expect(tsInterface).toMatchSnapshot()
    }))
  })

  describe('array of complex values', function() {
    const schema: Schema = {fields: {
      field: [new Complex({
        innerField: Number
      })]
    }}

    it('json-schema', testJSONSchema(schema, data => {
      expect(data).toMatchSnapshot()
    }))

    it('mongoose', testMongooseField(schema, 'field', field => {
      expect(field).toEqual({type: [{innerField: {type: Number}}]})
    }))

    it('typescript', testTypescriptInterface(schema, tsInterface => {
      expect(tsInterface).toMatchSnapshot()
    }))
  })
})
