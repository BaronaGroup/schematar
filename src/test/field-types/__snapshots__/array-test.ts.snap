// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array-test array of complex values json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"object\\",
        \\"required\\": [
          \\"innerField\\"
        ],
        \\"properties\\": {
          \\"innerField\\": {
            \\"type\\": \\"number\\"
          }
        },
        \\"additionalProperties\\": false
      }
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`array-test array of complex values typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType>
{
  field: Array<{
    innerField: number
  }>
}"
`;

exports[`array-test array of simple values defined with options json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"number\\"
      }
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`array-test array of simple values defined with options typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType>
{
  field: Array<number>
}"
`;

exports[`array-test array of simple values json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"number\\"
      }
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`array-test array of simple values typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType>
{
  field: Array<number>
}"
`;
